### DISCUSSION TITLE
RFC: Rich-text embedded variables

### DISCUSSION THREAD

**Discussion by jeffsee55** [tinacms/tinacms#4690] [Ideas]
## TLDR: 
We sometimes need to nest rich-text values in awkward areas, MDX and other markdown tools don't really give us what we need for it. We should store those values in a top-level "embeds" key, when we parse the file, we should transform the references back to their appropriate location:

```
---
title: Hello, world
author: 'authors/some-author.md'
embeds:
  abc: |
    Quote from [Somebody](https://example.com)

    and here is another paragraph!
---

<Quote description={embeds.abc}>
  # Hello, world!
</Quote>
```

The resulting payload wouldn't look any different than it currently does (expand to view)

<details>

```json
{
  "type": "root",
  "children": [
    {
      "type": "mdxJsxFlowElement",
      "name": "Quote",
      "children": [
        {
          "type": "text",
          "text": ""
        }
      ],
      "props": {
        "description": {
          "type": "root",
          "children": [
            {
              "type": "p",
              "children": [
                {
                  "type": "text",
                  "text": "Quote from "
                },
                {
                  "type": "a",
                  "url": "https://example.com",
                  "title": null,
                  "children": [
                    {
                      "type": "text",
                      "text": "Somebody"
                    }
                  ]
                }
              ]
            }
          ]
        },
        "children": {
          "type": "root",
          "children": [
            {
              "type": "h1",
              "children": [
                {
                  "type": "text",
                  "text": "Hello, world!"
                }
              ]
            }
          ]
        }
      }
    }
  ]
}
```
</details>

## Background

When working with MDX, we support nesting rich-text elements with some caveats. According to the MDX spec, only `children` will receive this special treatment. This is mostly fine for MDX users because it's trivial to nest children within children. For example

```
<Quote>
  # Hello world!

  <Description>
    Quote from [Somebody](https://example.com)
  </Description>
</Quote>
```

This solution requires that we overload the `children` value and would be a nightmare to manage from within the rich-text editor. So we have tried to support this need by inlining rich-text properties as normal props:

```
<Quote description={<>
Quote from [Somebody](https://example.com)
</>}>
  # Hello world!
</Quote>
```

This has been fine for most use-cases. The MDX parser treats the description prop as a javascript expression, so we're kind of cheating a bit and stripping the leading and trailing fragments, and then passing the remaining string back in to the parser. I've never loved this, and it's not been something we've seen a lot of usage for so hasn't been much of a priority. However, recently we had a user who wanted to do this with an inline component, which sort of broke our assumptions, because inline elements are not allowed to break onto new lines:

```
Hello there, here is my quote: <Quote description={<>Quote from [Somebody](https://example.com)</>}>Hello world!</Quote>, I hope you like it!
```

There's technically no reason that either the `children` prop or the `description` prop should _also_ need to be inline just because their parent is. The use case here is a UI that has a rich-text popover like this:

<img width="589" alt="Screenshot 2024-07-27 at 8 17 39 AM" src="https://github.com/user-attachments/assets/2050ceff-0bda-421c-9e17-ec5957f0f201">

You can imagine that the embed will have all sorts of data that doesn't adhere to being renderable in an inline element.

## Proposed solution
I'd like to introduce the concept of local embeds to rich-text. This would essentially pass  the document's frontmatter object to the rich-text parser and allow us to pop values off of a special "embeds" key:

```
---
title: Hello, world
author: 'authors/some-author.md'
embeds:
  abc: |
    Quote from [Somebody](https://example.com)

    and here is another paragraph!
---

<Quote description={embeds.abc}>
  # Hello, world!
</Quote>

```

**Advanced inline example**
The nice thing about this API is that inline or block-level can work the same as each other

```
---
title: Hello, world
author: 'authors/some-author.md'
embeds:
  abc: |
    A CMS stands for content management system.     <Definition id={embeds.def}>Wordpress</Definition> is probably the most popular one.
  def: |
    Wordpress is one of the most loved CMS solutions in existence. It's built with <Definition id={embeds.ghi}>PHP</Definition>
  ghi: |
    You get the idea
---

Tina is an open-source <Definition id={embeds.abc}>CMS</Definition> that works with basic files
```

Only `rich-text` fields would receive this treatment. If you have an object then any rich-text values within the structure will be treated as embeds

**Advanced object example**

```
<Table
  rows={[
  . {
      column1: embeds.abc
      column2: embeds.def
    },
    {
      column1: embeds.ghi
      column2: embeds.jkl
    },
  ]}
/>
```

This is, admittedly, a bit harder to read than if the values were kept very short.
#### Should inline rich-text children also have this behavior?
Right now when we parse `children` for an inline JSX element, we strip out the block-level piece and just return the first "paragraph" children nodes. It's kind of cryptic behavior, but I don't know that there's really anything wrong with this. If you want block-level rich-text in an inline node, just don't use `children` as the prop name. In general, whenever possible, I think we should try to keep the syntax the same as normal MDX expectations.

Eg.

```
It reads a bit better to just <Highlight>keep children inline</Highlight>
```
VS:
```
It reads a bit better to just <Highlight>{embeds.abc}</Highlight>
```

### Backwards compatability
As long as we don't delete the current logic, then this should just work™. The next time a document is saved throught Tina's API they will automatically be stored in the embeds. I don't think we'd want to support a way to opt out of this.

We _could_ document a way for users to make this changeover from the CLI, if there's enough of a demand for that.

### The "embeds" name
Open to suggestions here, we may want to make the name less likely to collide with user-land field definitions. I also don't have a ton of thoughts about the key naming, only that it should probably be random and maybe 6 letters long.

### Document-based context
Right now our parser and stringifier have no knowledge of any outside context. That is to say, when we parse, we don't know where in the document we're parsing, it's a potentially limitless recursive api. That would still be true, but we'll now need to carry around this global "embeds" context to read from during parse and populate during stringify. We already kind of have to deal with this with the `imageCallback`, which is responsible for prefixing the image url with the tina assets url.

### Key naming and ordering
The embed keys will need to be autogenerated, that might be a bit cryptic to readers who really care about the shape of their markdown. We may be able to offer some customization here, but won't be able to give them a callback. It's also important to note that they keys shouldn't represent order of any kind. And with that, deeply embedded embeds will also be referencing top-level embed keys, so best to just keep this structure opaque.

## Possible future work
I'd love to eventually turn this into what I would call "local variable" support. It'd be nice to be able to reference _any_ other value in the document for use with the rich-text editor. For example, you might want to embed the title of the document in your text. We might want to stick to other markdown tooling conventions and use `{{$frontmatter.title}}`, but I also wonder if we can just provide a component instead like `<Frontmatter key="title" />`. The latter would be a very small lift.

## Scope
This work is scoped mostly to changes within the `tinacms/mdx` package. And for now, only specific to the `mdx` parser. Those using shortcodes in markdown wouldn't get this behavior for now, but I don't see any reason we couldn't do it there too.

There shouldn't be much work on the rich-text editor itself. The only part that will need to be updated is the "raw" view. We'll need it to have access to the document's "embed" values, which they currently know nothing about. 

## Estimate
I would estimate 20 hours for work at this scope. I would love to pair up as much as possible and give ourselves a chance to share some knowledge!
